lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,3148 @@
+
+    auto id = defOp->getAttr("dnodeid").cast<IntegerAttr>().getUInt();
+    auto dnode = perOp[id];
+
+    if (!ifStack.empty() && defOp->getParentOp() != ifStack.back() &&
+        dnode->isQuantumDependent()) {
+      auto opdnode = static_cast<OpDependencyNode *>(dnode);
+      auto shadow_node = new ShadowDependencyNode{opdnode, resultidx};
+
+      auto parent = ifStack.back();
+
+      while (parent->getParentOp() != defOp->getParentOp())
+        parent = parent->getParentOp();
+
+      freeClassicals[parent].insert(shadow_node);
+
+      return DependencyNode::DependencyEdge{shadow_node, resultidx};
+    }
+
+    return DependencyNode::DependencyEdge{dnode, resultidx};
+  }
+
+  /// Cleans up semi-constructed dependency graph when backing out of running
+  /// DependencyAnalysis because of an encountered error
+  void clean() {
+    // TODO: clean up nodes
+    // Adam: can use perOps, have to be careful about nested nodes though
+  }
+
+  unsigned getNumVirtualAllocs() { return vallocs; }
+};
+
+struct DependencyAnalysisPass
+    : public cudaq::opt::impl::DependencyAnalysisBase<DependencyAnalysisPass> {
+  using DependencyAnalysisBase::DependencyAnalysisBase;
+
+  /// DependencyAnalysis constructs a data structure representing the
+  /// quake AST, performs several analyses/optimizations, and then generates
+  /// a new quake AST based on the resulting data structure.
+  ///
+  /// First, the quake AST is walked, constructing a DependencyBlock for the
+  /// body of every kernel function.
+  ///
+  /// Next, virtual qubit allocations are lowered to the inner-most scope in
+  /// which they are used (see `contractAllocsPass`). This works outside-in.
+  /// Lowering virtual qubit allocations opens up more qubit reuse opportunities
+  /// within inner scopes, and gives the remaining optimizations more
+  /// flexibility.
+  ///
+  /// Then, an inside-out analysis/optimization pass is performed, assigning
+  /// physical qubits to virtual wires, and lifting common optimizations. This
+  /// inside-out analysis/optimization pass works as follows:
+  /// - Step 1: `if`s inside blocks are resolved (hence inside-out), once an
+  /// `if` is resolved, the information contained by it will not be changed
+  /// (with the possible exception of re-indexing physical qubits).
+  /// - Step 2: blocks are resolved.
+  /// * First the nodes inside them are scheduled, assigning a cycle to every
+  ///   node such that each node is scheduled after all of its dependencies, and
+  ///   before all of its successors.
+  /// * Then, physical qubits are allocated and assigned to virtual wires based
+  ///   on lifetime information (i.e., which cycles the virtual wire is used
+  ///   in). This algorithm treats `if`s as "solid rectangles", where all qubits
+  ///   in use anywhere in the `if` are considered in use for the entire `if` by
+  ///   the parent scope. In other words, the lifetime analysis does not look
+  ///   inside `if`s.
```

Comment:
This considers qubits that are temporally live _only_ within the body of the `if` as "used by the if", right?

Abstractly, that is the biggest difference between an `if` macro operation and a slimmer quake operation. But necessarily imply that their quantum uses are live-in and their quantum outputs are live-out. The "bonus" use concept might be useful for operations that imply ancilla qubits as well.

Reply:
Right, any qubit used anywhere in the `if` is considered "in use" for the whole `if`, whether it is passed as a linear argument or only allocated in the body.